LCOW: Support for docker cp, ADD/COPY on build

gupta-ak opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/docker/docker/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**

Added support for LCOW container file operations like`docker cp` and `ADD/COPY` for `docker build`

**- How I did it**

1. The original graphdriver interface for `Get()` returned a `(string, error)`. I changed it to return a new interface `RootFS` defined in `pkg/rootfs/rootfs.go`.  It has the `Driver` and `PathDriver` interface from `continuity`, which abstracts the file operations from the `os` package behind the interface. The `RootFS` interface also has container specific methods, such as resolving scoped paths.
2. I implemented the `RootFS` interface for non-LCOW and LCOW cases. In the non-LCOW cases, it's just a wrapper around the `os` package, but for LCOW, docker sends the file operations to a Linux Service VM that handles it.
3. After that, it was a matter of propagating the interface throughout docker.

**- How to verify it**

For regressions due to the interface break, I'm hoping the CI will catch them. For LCOW, I'm running @jhowardmsft 's LCOW testing script and also manually verifying that `docker cp` and `docker build` works.

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->

Support for docker cp and ADD/COPY in docker build

**- A picture of a cute animal (not mandatory but encouraged)**

![image](https://user-images.githubusercontent.com/13004234/28589606-af7aab06-7133-11e7-9f1f-176728c9accb.png)

gupta-ak commented:
Example of docker cp working:
```
PS C:\Users\Administrator> cat '.\Documents\New folder\Dockerfile'
FROM busybox
ENV "Goldens" "Are the best dogs"
RUN export
#Hello!
PS C:\Users\Administrator> docker create -it busybox
8691c852d3bfc68201db40183d875bed2f3cba58174e30f429ef2811a12ce912
PS C:\Users\Administrator> docker cp '.\Documents\New folder\Dockerfile' 869:/root/
PS C:\Users\Administrator> docker start -ia 869
/ # cat /root/Dockerfile
FROM busybox
ENV "Goldens" "Are the best dogs"
RUN export
#Hello!/ #exit
PS C:\Users\Administrator> docker cp 869:/root/Dockerfile Dockerfile2
PS C:\Users\Administrator> cat .\Dockerfile2
FROM busybox
ENV "Goldens" "Are the best dogs"
RUN export
#Hello
PS C:\Users\Administrator>
```

gupta-ak commented:
@stevvooe PTAL as well. Right now, everything is complete aside from the lcow/remotefs implementation, which mostly works, but has some reliability issues.

This PR can be split into multiple PRs (one for interface changes, one for LCOW changse, one for remotefs implementation) so if that is easier to do, I can do that as well.

dnephin commented on builder/dockerfile/internals.go:
```diff
@@ -298,3 +340,39 @@ func hostConfigFromOptions(options *types.ImageBuildOptions) *container.HostConf
+	return false
+}
+
+func validateCopySourcePath(imageSource *imageMount, origPath, platform string) error {
```
Please move these two helpers to below the single function where they are called.

dnephin commented on builder/dockerfile/internals.go:
```diff
@@ -298,3 +340,39 @@ func hostConfigFromOptions(options *types.ImageBuildOptions) *container.HostConf
+
+func validateCopySourcePath(imageSource *imageMount, origPath, platform string) error {
+	// validate windows paths from other images
+	if imageSource == nil || platform != "windows" {
```
A type and constant for platform and "windows" would be nice

gupta-ak replied:
There's quite a few `platform string` that was already here. I think it's from @jhowardmsft's first LCOW PR. He's on a platform API for the builder, so do you want the change now or in his PR? @jhowardmsft: Thoughts?

jhowardmsft replied:
I'm not sure this makes sense at this point*. There's dozens of places outside of LCOW where platform strings are used directly without a constant. It's been a long-standing pattern like this.

```
E:\go\src\github.com\docker\docker [jjh/buildplatformapi]> (git grep '\"windows\"' *.go | Measure-Object).Count
262
```

\* Sure, no objection to doing this a little further down the road, across the entire code base (ideally at a point where the fast-arriving LCOW PRs aren't going to merge conflict).

dnephin commented on builder/dockerfile/copy.go:
```diff
@@ -82,14 +83,20 @@ func (o *copier) createCopyInstruction(args []string, cmdName string) (copyInstr
+		// Work in daemon-specific filepath semantics
+		inst.dest = filepath.FromSlash(args[last])
+		separator = string(os.PathSeparator)
+	}
```
For the most part the path details seem to be well abstracted, but they leak out in this function.

Instead of this maybe we could have a `system.HasDirectorySuffix()` and `system.FromSlash()` or something like that?

`system` is already a large package, so maybe that's not the right place for it.

gupta-ak replied:
I added a platform tag to the `copier` struct. The platform was already in the `dispathRequest.builder`, so I just had to extract that into the`copierFromDistpatchRequest`. I added a `system.FromSlash()` and `system.Separator()` to check the directory suffix. 

If the copy instruction `dest` doesn't need the daemon/platform specific paths here, we could also just do a check like `if len(infos) > 1 && !strings.HasSuffix(filepath.ToSlash(inst.dest), "/")`

gupta-ak commented:
Couple of updates.

- I squashed all the commits to a single one & vendor commit. 
- Finished rest of remotefs implementation, so the docker code should be ready. There's some platform reliability issues I'm hitting when testing this that I'm investigating, but the docker code shouldn't change.
- Addressed @dnephin's comments

gupta-ak committed:
LCOW: Implemented support for docker cp + build

This enables docker cp and ADD/COPY docker build support for LCOW.
Originally, the graphdriver.Get() interface returned a local path
to the container root filesystem. This does not work for LCOW, so
the Get() method now returns an interface that LCOW implements to
support copying to and from the container.

Signed-off-by: Akash Gupta <akagup@microsoft.com>

gupta-ak committed:
Vendor containerd/continuity@22694c680ee48fb8f50015b44618517e2bde77e8

Signed-off-by: Akash Gupta <akagup@microsoft.com>

gupta-ak commented:
Now that #34170 is merged. I rebased to just my commit and the vendor.

jhowardmsft commented on daemon/graphdriver/lcow/lcow_svm.go:
```diff
@@ -0,0 +1,394 @@
+	"io"
+
+	"github.com/Microsoft/hcsshim"
+	"github.com/jhowardmsft/opengcs/gogcs/client"
```
This should switch over to gitHub.com/Microsoft/opengcs/client now.

jhowardmsft commented on daemon/graphdriver/lcow/lcow.go:
```diff
@@ -349,133 +320,115 @@ func (d *Driver) startServiceVMIfNotRunning(id string, mvdToAdd *hcsshim.MappedV
-		if err := svm.config.Uvm.Terminate(); err != nil {
-			return fmt.Errorf("failed to terminate utility VM (%s): %s", context, err)
+	logrus.Debugf("lcowdriver: terminateservicevm: %s (%s) - calling terminate", id, context)
+	if err := svm.config.Uvm.Terminate(); err != nil {
```
This is problematic. I hit a nil pointer de-reference panic due to this change at L411. Previously, the code had an explicit nil check here. Repro was docker cp IDOfStoppedContainer:/bin/sh .

gupta-ak replied:
I can't repro it, but the logic in `startServiceVMIfNotRunning` needs some tweaking. See here:
```go
defer func() {
	// Signal that start has finished, passing in the error if any.
	svm.signalStartFinished(err)
	if err == nil {
		return
	}

	// We added a ref to the VM, since we failed, we should delete the ref.
	d.terminateServiceVM(id, "error path on startServiceVMIfNotRunning", false)
}()

/* Some functions that can fail. */
	
// Start it.
logrus.Debugf("lcowdriver: startServiceVmIfNotRunning: (%s) starting %s", context, svm.config.Name)
if err := svm.config.StartUtilityVM(); err != nil {
	return nil, fmt.Errorf("failed to start service utility VM (%s): %s", context, err)
}
```
Bascially, if any function fails before the `svm.config.StartUtilityVM()` part, then `svm.config.Uvm` will be nil, triggering a panic, so I need to either guard L411 with some check.

jhowardmsft commented on daemon/graphdriver/lcow/lcow.go:
```diff
@@ -349,133 +320,115 @@ func (d *Driver) startServiceVMIfNotRunning(id string, mvdToAdd *hcsshim.MappedV
-			return fmt.Errorf("failed to terminate utility VM (%s): %s", context, err)
+	logrus.Debugf("lcowdriver: terminateservicevm: %s (%s) - calling terminate", id, context)
+	if err := svm.config.Uvm.Terminate(); err != nil {
+		// We might get Operationg still pending from the HCS. In that case, we shouldn't return
```
s/Operationg/operation

gupta-ak commented:
@stevvooe PTAL + add whoever else should take a look

jhowardmsft commented on daemon/graphdriver/lcow/lcow_svm.go:
```diff
@@ -0,0 +1,394 @@
+	stopStatus chan interface{}
+	stopError  error
+
+	// NOTE: It is OK to use a cache here because Windows does not support
```
Has this comment been moved to the correct place? It used to be in the driver struct for the cache.

jhowardmsft commented on daemon/graphdriver/lcow/lcow_svm.go:
```diff
@@ -0,0 +1,394 @@
+	scratchAttached bool           // Has a scratch been attached?
+	config          *client.Config // Represents the service VM item.
+
+	// Indicate that the vm is started
```
Typo: Indicates

jhowardmsft commented on daemon/graphdriver/lcow/lcow_svm.go:
```diff
@@ -0,0 +1,394 @@
+	}
+}
+
+// add will add an id to the service vm map. There are a couple of cases:
```
three cases ;)

jhowardmsft commented on daemon/graphdriver/lcow/lcow_svm.go:
```diff
@@ -0,0 +1,394 @@
+
+// serviceVMWithRef is our internal structure representing an item in our
+// map of service VMs we are maintaining.
+type serviceVMWithRef struct {
```
More nit than anything else, I was kind of confused to start with by the name here. Perhaps `serviceVMMapItem` would be clearer?

jhowardmsft commented on daemon/graphdriver/lcow/lcow_svm.go:
```diff
@@ -0,0 +1,394 @@
+	unionMounts  map[string]int // Map ref counting all the union filesystems we mounted.
+}
+
+func newServiceVMMap() *serviceVMMap {
```
Wondering if it's worth having this constructor rather than inlining in driver init?

jhowardmsft commented on daemon/graphdriver/lcow/lcow_svm.go:
```diff
@@ -0,0 +1,394 @@
+// 	- entry doesn't exist:
+// 		- add id to map and return a new vm that the caller can manually configure+start
+//	- entry does exist
+//  	- return vm in map and add to ref count
```
s/add to/increment

jhowardmsft commented on daemon/graphdriver/lcow/lcow_svm.go:
```diff
@@ -0,0 +1,394 @@
+	}
+
+	// Doesn't exist, so create an empty svm to put into map and return
+	newSVM := newServiceVM()
```
Similar to comment to previously - do we really need this simple constructor function. Looks like it's only called from here. Inline?

jhowardmsft commented on daemon/graphdriver/lcow/lcow_svm.go:
```diff
@@ -0,0 +1,394 @@
+	return newSVM, false, nil
+}
+
+// get will get the service vm from the map. There are a couple of cases:
```
Three cases here too :)

jhowardmsft commented on daemon/graphdriver/lcow/lcow_svm.go:
```diff
@@ -0,0 +1,394 @@
+	return svm.svm, nil
+}
+
+// Reduces the ref count of the given ID from the map. There are a couple of cases:
```
Strictly the comment should start `reduceRef....`. And `decrements the...` rather than `reduces the`? Maybe the function should just be called `decrementRefCount`?

jhowardmsft commented on daemon/graphdriver/lcow/lcow.go:
```diff
@@ -349,133 +320,115 @@ func (d *Driver) startServiceVMIfNotRunning(id string, mvdToAdd *hcsshim.MappedV
-			if err := os.Remove(scratchTargetFile); err != nil {
-				logrus.Warnf("failed to remove scratch file %s (%s): %s", scratchTargetFile, context, err)
-			}
+		scratchTargetFile := filepath.Join(d.dataRoot, scratchDirectory, fmt.Sprintf("%s.vhdx", id))
```
Why has the `if svm.scratchAttached {` check been removed?

gupta-ak replied:
Yeah, there's a bug similar to https://github.com/moby/moby/pull/34252#discussion_r132332945 here without the `if svm.scratchAttached {` check.

jhowardmsft commented on daemon/graphdriver/lcow/lcow_svm.go:
```diff
@@ -0,0 +1,394 @@
+	return svm.svm, nil
+}
+
+// Deletes the given ID from the map. If the refcount is not 0 or the
```
Again, strictly the comment should start `deleteID`. Nit as not exported though.

jhowardmsft commented on daemon/graphdriver/lcow/lcow.go:
```diff
@@ -349,133 +320,115 @@ func (d *Driver) startServiceVMIfNotRunning(id string, mvdToAdd *hcsshim.MappedV
-		logrus.Debugf("%s (%s): releasing serviceVM", title, context)
-		svm.Unlock()
 	}
 
```
The debug statement is incorrect here now as the lock isn't taken.

jhowardmsft commented on daemon/graphdriver/lcow/lcow.go:
```diff
@@ -349,133 +320,115 @@ func (d *Driver) startServiceVMIfNotRunning(id string, mvdToAdd *hcsshim.MappedV
 	logrus.Debugf("%s locking cachedScratchMutex", title)
 	d.cachedScratchMutex.Lock()
 	if _, err := os.Stat(d.cachedScratchFile); err != nil {
 		logrus.Debugf("%s (%s): creating an SVM scratch - locking serviceVM", title, context)
```
urm the lock isn't being taken any more....

jhowardmsft commented on daemon/graphdriver/lcow/lcow_svm.go:
```diff
@@ -0,0 +1,394 @@
+	return svm.hotAddVHDsNoLock(mvds...)
+}
+
+func (svm *serviceVM) hotAddVHDsNoLock(mvds ...hcsshim.MappedVirtualDisk) error {
```
Maybe stylistic, but I would prefer just a `hotAddVHDs` function (same for remove), and no `...NoLock()` function. Then  put the onus on the caller to ensure the lock is taken/defer-unlocked. And then just add a comment before `hotAddVHDs saying who is responsible for locking`.

jhowardmsft commented on daemon/graphdriver/lcow/lcow_svm.go:
```diff
@@ -0,0 +1,394 @@
+func (svm *serviceVM) hotRemoveVHDsNoLock(mvds ...hcsshim.MappedVirtualDisk) error {
+	var retErr error
+	for _, mvd := range mvds {
+		hostPath := mvd.HostPath
```
Don't need this variable. Just use `mvd.HostPath`, the same as for the matching HotAdd function

jhowardmsft commented on daemon/graphdriver/lcow/lcow_svm.go:
```diff
@@ -0,0 +1,394 @@
+		hostPath := mvd.HostPath
+		_, ok := svm.attachedVHDs[hostPath]
+		if !ok {
+			continue
```
Really a continue? Or an error?  

General comment - it might be useful to have some debug logging here to indicate what's being removed (unless it's covered already by opengcs which I suspect it is)

gupta-ak replied:
The thing is that `Get` defers the VM start, so it's possible that the VM hasn't even been created & nothing has been hot-added before `Put` is called. For example, the continue was changed to a return error, then anything like this:
```go
// Just want the container mount point
mnt, err := driver.Get(id, mntName)
err = driver.Put(id)
```
will return an error for `driver.Put()` which seems odd to me. I could return an error here and then handle the error higher up in `Put()`,  but since these helper functions are used exclusively in the lcow driver, it's effectively the same thing.

gupta-ak replied:
Btw, the logging is in the opengcs, but I can add logging for the refcount.

jhowardmsft replied:
Got it. Maybe add a comment to clarify this though? (I like comments :) )

jhowardmsft commented on daemon/graphdriver/lcow/lcow_svm.go:
```diff
@@ -0,0 +1,394 @@
+			continue
+		}
+
+		if svm.attachedVHDs[hostPath] != 1 {
```
`> 0` seems a tiny bit clearer. Just a nit.

jhowardmsft commented on daemon/graphdriver/lcow/lcow_svm.go:
```diff
@@ -0,0 +1,394 @@
+
+	// NOTE: It is OK to use a cache here because Windows does not support
+	// restoring containers when the daemon dies.
+	attachedVHDs map[string]int // Map ref counting all the VHDS we've mounted/unmounted.
```
`we've hot-added/hot-removed` rather than `mounted/unmounted`?

jhowardmsft commented on daemon/graphdriver/lcow/lcow_svm.go:
```diff
@@ -0,0 +1,394 @@
+
+func (svm *serviceVM) createUnionMount(mountName string, mvds ...hcsshim.MappedVirtualDisk) (err error) {
+	if len(mvds) == 0 {
+		return fmt.Errorf("createUnionMount: error must have atleast 1 layer")
```
typo: at least

jhowardmsft commented on daemon/graphdriver/lcow/lcow_svm.go:
```diff
@@ -0,0 +1,394 @@
+
+	svm.Lock()
+	defer svm.Unlock()
+	_, ok := svm.unionMounts[mountName]
```
`if _, ok := svm.unionMounts[mountName]; ok {` ....?

jhowardmsft commented on daemon/graphdriver/lcow/lcow_svm.go:
```diff
@@ -0,0 +1,394 @@
+	}
+
+	logrus.Debugf("Doing the overlay mount with union directory=%s", mountName)
+	err = svm.runProcess(fmt.Sprintf("mkdir -p %s", mountName), nil, nil, nil)
```
Can you make sure mkdir is listed as a required utility in the opengcs initrd build instructions?

gupta-ak replied:
Will do for this and mount

gupta-ak replied:
nevermind, the gcs already uses them, so it's not needed.

jhowardmsft commented on daemon/graphdriver/lcow/lcow_svm.go:
```diff
@@ -0,0 +1,394 @@
+	var cmd string
+	if mvds[0].ReadOnly {
+		// Readonly overlay
+		cmd = fmt.Sprintf("mount -t overlay overlay -olowerdir=%s %s",
```
Same to make sure /bin/mount is now a required utility

jhowardmsft commented on daemon/graphdriver/lcow/lcow_svm.go:
```diff
@@ -0,0 +1,394 @@
+		upper := fmt.Sprintf("%s/upper", mvds[0].ContainerPath)
+		work := fmt.Sprintf("%s/work", mvds[0].ContainerPath)
+
+		err = svm.runProcess(fmt.Sprintf("mkdir -p %s %s", upper, work), nil, nil, nil)
```
`if err:= .... ; err != nil {`

jhowardmsft commented on daemon/graphdriver/lcow/lcow_svm.go:
```diff
@@ -0,0 +1,394 @@
+	}
+
+	logrus.Debugf("createUnionMount: Executing mount=%s", cmd)
+	err = svm.runProcess(cmd, nil, nil, nil)
```
`if err := ...; err != nil {`

gupta-ak replied:
ok, I'll go find them

jhowardmsft commented on daemon/graphdriver/lcow/lcow_svm.go:
```diff
@@ -0,0 +1,394 @@
+		return nil
+	}
+
+	if svm.unionMounts[mountName] != 1 {
```
> 0 rather than != 1

jhowardmsft commented on daemon/graphdriver/lcow/lcow.go:
```diff
@@ -106,72 +109,22 @@ const (
-	scratchAttached bool           // Has a scratch been attached?
-	config          *client.Config // Represents the service VM item.
-}
+	// ErrOperationPending is the HRESULT returned by the HCS when the VM termination operation is still pending.
```
lowercase e on Err

jhowardmsft commented on pkg/rootfs/rootfs.go:
```diff
@@ -0,0 +1,79 @@
+type Driver interface {
+	// Platform returns the OS where the rootfs is located. Essentially,
+	// runtime.GOOS for everything aside from LCOW, which is "linux"
+	Platform() string
```
Just wondering how this is used and whether it's needed? Still trying to figure that out.

gupta-ak replied:
It's usually for `OS` specific path manipulation that can't be done with the normal filepath functions, because Windows Containers have specific requirements or do things like add `C:\`. For example here in `builder/dockerfile/copy.go` :

```go
func validateCopySourcePath(imageSource *imageMount, origPath, platform string) error {
	// validate windows paths from other images
	if imageSource == nil || platform != "windows" {
		return nil
	}

	origPath = filepath.FromSlash(origPath)
	p := strings.ToLower(filepath.Clean(origPath))
	if !filepath.IsAbs(p) {
		if filepath.VolumeName(p) != "" {
			if p[len(p)-2:] == ":." { // case where clean returns weird c:. paths
				p = p[:len(p)-1]
			}
			p += "\\"
		} else {
			p = filepath.Join("c:\\", p)
		}
	}
	if _, blacklisted := pathBlacklist[p]; blacklisted {
		return errors.New("copy from c:\\ or c:\\windows is not allowed on windows")
	}
	return nil
}
```

jhowardmsft commented on daemon/graphdriver/lcow/remotefs_filedriver.go:
```diff
@@ -0,0 +1,248 @@
+	"github.com/sirupsen/logrus"
+)
+
+// TODO: @gupta-ak. These constants are defined in the Microsoft/opengcs repo.
```
This can be done now :)

jhowardmsft commented on daemon/graphdriver/lcow/remotefs_filedriver.go:
```diff
@@ -0,0 +1,248 @@
+
+	// After OpenFile, we know that the file exists in the VM.
+	// If it's a regular file, retrieve it
+	var file *os.File
```
Don't think you need the var here. Instead just have `file, err := ioutil......` in the if and else blocks.

jhowardmsft replied:
Ah, ignore - I see it's used lower down too.

jhowardmsft commented on daemon/graphdriver/lcow/remotefs_filedriver.go:
```diff
@@ -0,0 +1,248 @@
+		}
+
+		// We do this so that we can match the flags that were given to us.
+		file.Close()
```
I'm a bit confused by this and what this is doing. Can you explain? Also, AFAIK, I don't think the flag/perm are implemented in the Windows go libraries, so I'm not sure it would have any effect.

gupta-ak replied:
Oh I didn't know the permission bits weren't implemented on Windows.  

Anyway, I am reworking the implementation for this. Originally, I had the file copied to the host and the read/writes/seeks would happen on the host. On Close, the file would be sent to the service VM. It was a hack to avoid keeping state in my `remotefs` program.  I realized I could just use the `remotefs` process as a handle to a single file and have a simple protocol to forward reads/writes/seeks/closes to it, so I will change this side of the code to reflect that. The updated `OpenFile` on the gcs side is here: https://github.com/Microsoft/opengcs/pull/94

jhowardmsft commented on daemon/graphdriver/lcow/remotefs_filedriver.go:
```diff
@@ -0,0 +1,248 @@
+	}, nil
+}
+
+func writeAllowed(flag int) bool {
```
Seems pretty heavy-weight for a single use....

jhowardmsft commented on daemon/graphdriver/lcow/remotefs_filedriver.go:
```diff
@@ -0,0 +1,248 @@
+	return &fi, nil
+}
+
+// TODO: @gupta-ak. Remove this struct once the switch to Mircrosoft/opengcs is done.
```
This can be done too now :)

jhowardmsft commented on daemon/graphdriver/lcow/remotefs_pathdriver.go:
```diff
@@ -0,0 +1,212 @@
+import (
+	"errors"
+	"os"
+	pathpkg "path"
```
Why aliased to pathpkg and imported as `path` as well a few lines below?

gupta-ak replied:
oops, guess my editor automatically added the `path` import

jhowardmsft replied:
Yeah, but why the alias?

gupta-ak replied:
The `path` variables in the function headers like `func (l *lcowfs) FromSlash(path string) string ` shadow the `path` package, so I aliased the package. I could also change the path variables in the function headers, but I wanted them to match the ones you see in `path/filepath`

jhowardmsft commented on daemon/graphdriver/lcow/remotefs_pathdriver.go:
```diff
@@ -0,0 +1,212 @@
+}
+
+func (l *lcowfs) FromSlash(path string) string {
+	return path
```
Probably worth adding a comment to say that the go standard library posix From/ToSlash functions are no-ops in the case that the path separator is `/`. I had to remind myself to check.

jhowardmsft commented:
From what I've reviewed so far (mainly focused on the lcow driver changes, I've only skimmed the rest a couple of times), it's looking pretty good, most of my comments so far are nits. I'll find time to review the rest of the changes hopefully tomorrow.

@johnstep Can you also review the graphdriver bit of changes?

stevvooe commented on pkg/rootfs/utils.go:
```diff
@@ -0,0 +1,61 @@
+// continuity interface
+
+// ReadFile works the same as ioutil.ReadFile with the rootFS abstraction
+func ReadFile(r RootFS, filename string) ([]byte, error) {
```
Could we move these to continuity?

gupta-ak commented:
rebased + addressed @jhowardmsft  feedback.

dmcgowan commented on daemon/graphdriver/driver.go:
```diff
@@ -112,6 +113,13 @@ type Driver interface {
 	DiffDriver
 }
 
+// LayerGetter is an interface to get the path to a layer on the host.
```
This could just be defined in `layer/layer_windows.go` if that is the only place it is used

dmcgowan commented:
Is there any benchmarks that have been done on this vs some of the other drivers? I am curious about the performance of push since it appears to be relying on tar split which require remotely reading all the file content for a layer on push.

dmcgowan commented on pkg/archive/copy.go:
```diff
@@ -103,22 +106,34 @@ func TarResourceRebase(sourcePath, rebaseName string) (content io.ReadCloser, er
+
+// TarResourceRebaseOpts does not preform the Tar, but instead just creates the parameters
+// to be sent to TarWithOptions (the source directory and the TarOptions struct)
+func TarResourceRebaseOpts(sourcePath string, rebaseName string, driver pathdriver.PathDriver) (string, *TarOptions) {
```
Can we have this function just take in the `sourceDir` and `sourceBase`? That would eliminate the need to pass in the `pathdriver` here and changing the `SplitPathDirEntry`. I would like to eliminate all use of the `pathdriver` in the archive package if possible.

gupta-ak replied:
Yeah it's possible to do that and have a helper function elsewhere that does the split with the `pathdriver` interface.

gupta-ak replied:
The last reference to the `pathdriver` is `archive.PreserveTrailingDotOrSeparator`. That's used by `container/archive.go` and `daemon/archive.go`. It's always paired with `system.CheckSystemDriveAndRemoveDriveLetter`, so maybe the container struct can provide a method that does those two like `CleanContainerPath` and then we can remove the `pathdriver` reference from those two functions.

gupta-ak commented:
@dmcgowan I rebased to master and removed the continuity pathdriver dependency on the archive pkg.

dmcgowan commented on pkg/system/path.go:
```diff
@@ -19,3 +26,51 @@ func DefaultPathEnv(platform string) string {
+}
+
+// FromSlash works like filepath.FromSlash but with a given OS platform field
+func FromSlash(path, platform string) string {
```
Why is this one handled differently in regards to LCOW than the previous function?

gupta-ak replied:
It's used in the builder for setting up the copy instructions, where you don't have a containerfs abstraction to call the `containerfs.FromSlash`  function. Since it's only used in the builder, I can remove these methods from the system package and add them as unexported methods to the builder.

dmcgowan commented on pkg/containerfs/archiver.go:
```diff
@@ -0,0 +1,195 @@
+	// as owner
+	rootIDs := archiver.IDMappingsVar.RootPair()
+	// Create dst, copy src's content into it
+	logrus.Debugf("Creating dest directory: %s", dst)
```
This debug line doesn't seem necessary, there is already a debug line around calling tar/untar. If the `err` does not provide enough context I would suggest using `errors.Wrapf`

gupta-ak replied:
The Archiver code for `CopyWithTar` in `pkg/archive` has the same logging statements in `moby/master`. Should I remove it there as well?

dmcgowan commented on pkg/containerfs/archiver.go:
```diff
@@ -0,0 +1,195 @@
+// CopyFileWithTar emulates the behavior of the 'cp' command-line
+// for a single file. It copies a regular file from path `src` to
+// path `dst`, and preserves all its metadata.
+func (archiver *Archiver) CopyFileWithTar(src, dst string) (err error) {
```
This function is so bizarre. This is out of scope for your change since you are preserving existing behavior, but were you able to fully understand why this is necessary? I understand the directory copy with tar, but this seems so heavy.

gupta-ak replied:
I'm not exactly sure why this method exists either. It's only used for `copyFile` in the `builder\dockerfile\copy.go`, where it has a `chrootarchive.Untar` as the plugin untar function, and `makePluginBundle` in `intergration-cli\fixtures\plugin`, where it has the `archive.Untar` as the plugin untar function. It seems like it could be removed with just a regular copyfile with an option for a chroot copyfile.

dmcgowan commented on daemon/graphdriver/fsdiff.go:
```diff
@@ -35,6 +35,7 @@ type NaiveDiffDriver struct {
 //     Changes(id, parent string) ([]archive.Change, error)
 //     ApplyDiff(id, parent string, diff archive.Reader) (size int64, err error)
 //     DiffSize(id, parent string) (size int64, err error)
+// Note that these assume a local file system for the graph driver (does not support LCOW)
```
I would just mention in the description that the `NaiveDiffDriver` operates by calculating and applying diffs using the local file system. Using the naive diff is a choice of the graph driver implementer, no need to mention LCOW.

dmcgowan commented on daemon/graphdriver/driver.go:
```diff
@@ -152,6 +153,20 @@ type Checker interface {
+
+// WrapLocalGetFunc wraps the old graphdriver Get() interface (LocalGetFunc)
+// with the current graphdriver.Get() interface
+func WrapLocalGetFunc(id, mountLabel string, f LocalGetFunc) (containerfs.ContainerFS, error) {
```
It might make the change a little bigger, but I would prefer to see the graph drivers just use `containerfs.NewLocalContainerFS` and not this wrap method. This adds an extra method call and is kind of complex, just returning nil or `containerfs.NewLocalContainerFS` is simple enough.

dmcgowan commented on builder/dockerfile/copy.go:
```diff
@@ -200,27 +209,64 @@ func (o *copier) calcCopyInfo(origPath string, allowWildcards bool) ([]copyInfo,
 	return newCopyInfos(newCopyInfoFromSource(o.source, origPath, hash)), nil
 }
 
+func validateCopySourcePath(imageSource *imageMount, origPath, platform string) error {
```
Did this need to be moved out of the windows only file? I understand some linux functionality needs to be moved so it can be used with Windows builds, but the other direction shouldn't be needed (unless we are doing WCOL 😱)

gupta-ak replied:
I see. I can revert it back to a Windows implemenation for LCOW/WCOW and a no-op Unix implementation.

gupta-ak commented:
Updated with @dmcgowan's feedback

johnstep commented:
I'm still reviewing the graph driver changes, and plan to finish by Monday. Apologies for taking longer than expected.

johnstep commented on daemon/graphdriver/lcow/lcow_svm.go:
```diff
@@ -0,0 +1,374 @@
+	}
+
+	if svm.unionMounts[mountName] > 0 {
+		svm.unionMounts[mountName]--
```
Shouldn't this function remove the mount if this is the last reference?

gupta-ak replied:
good catch!

johnstep commented on daemon/graphdriver/lcow/lcow_svm.go:
```diff
@@ -0,0 +1,374 @@
+	"sync"
+	"time"
+
+	"io"
```
Nit: Why is `io` separate from the previous set of imports?

gupta-ak replied:
seems like a typo

johnstep reviewed (approved):
I only skimmed the rest of the PR, but the LCOW graph driver changes look good!

mlaventure commented:
I just got aware of this, and it's too big for me to have a proper look in a decent timeframe, but it does look like overkill. I think something is wrong if we have to re-implement the POSIX interface within moby. This would be better served by an actual filesystem driver. Any reason not to go that way? It would also allow Microsoft to change its behavior in sync with their shim without requiring changes within Docker.

gupta-ak commented on daemon/graphdriver/lcow/lcow.go:
```diff
@@ -67,10 +67,13 @@ import (
 	"sync"
 	"time"
 
+	"syscall"
```
just seems like a typo

gupta-ak commented:
rebased + updated with @johnstep 's feedback

vieux committed:
Add LCOW behind experimental,
might not be the cleanest way, but it's definitly the way with the
minimum code change.

Signed-off-by: Victor Vieux <victorvieux@gmail.com>

johnstep reviewed (approved):
LGTM (based on code, will test soon)

jhowardmsft commented:
Graphdriver changes LGTM

thaJeztah commented:
from https://github.com/moby/moby/pull/34252#issuecomment-325053250

> Notes from 8/25 call with Li Li, Akash, Derek, Tonis, John S, Friis, Carl

> Next steps:
> 1 - Get 2 LGTMs on LCOW portions (@jhowardmsft, @johnstep )
> 2 - Get 1 LGTM from Docker on the non-LCOW portions (@dmcgowan)
> 3 - MSFT to work with Docker (@friism to get contact) to get graph driver test coverage via existing > end to end regression test suite

@dmcgowan @mlaventure is this one ready to go?

dmcgowan commented:
LGTM
